service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,318 @@
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal arguments: %w", err)
+	}
+	return getHash(b), nil
+}
+
+// New returns a new instance of the remotecfg service.
+func New(opts Options) (*Service, error) {
+	basePath := filepath.Join(opts.StoragePath, ServiceName)
+	err := os.MkdirAll(basePath, 0750)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Service{
+		opts:   opts,
+		ticker: time.NewTicker(math.MaxInt64),
+	}, nil
+}
+
+// Data is a no-op for the remotecfg service.
+func (s *Service) Data() any {
+	return nil
+}
+
+// Definition returns the definition of the remotecfg service.
+func (s *Service) Definition() service.Definition {
+	return service.Definition{
+		Name:       ServiceName,
+		ConfigType: Arguments{},
+		DependsOn:  nil, // remotecfg has no dependencies.
+	}
+}
+
+var _ service.Service = (*Service)(nil)
+
+// Run implements [service.Service] and starts the remotecfg service. It will
+// run until the provided context is canceled or there is a fatal error.
+func (s *Service) Run(ctx context.Context, host service.Host) error {
+	s.ctrl = host.NewController(ServiceName)
+
+	s.fetch()
+
+	// Run the service's own controller.
+	go func() {
+		s.ctrl.Run(ctx)
+	}()
+
+	for {
+		select {
+		case <-s.ch:
+			b, err := s.getAPIConfig()
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to fetch configuration from the API", "err", err)
+				continue
+			}
+
+			// The polling loop got the same configuration, no need to reload.
+			newConfigHash := getHash(b)
+			if s.getCfgHash() == newConfigHash {
+				level.Debug(s.opts.Logger).Log("msg", "skipping to the next polling loop")
+				continue
+			}
+
+			// We have a new configuration let's try to load it
+			err = s.parseAndLoad(b)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to load configuration from the API", "err", err)
+				continue
+			}
+
+			level.Info(s.opts.Logger).Log("msg", "new remote configuration loaded successfully")
+
+			// If successful, flush to disk and keep a copy.
+			err = os.WriteFile(s.dataPath, b, 0750)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to flush remote configuration contents the on-disk cache", "err", err)
+			}
+			s.setCfgHash(newConfigHash)
+		case <-ctx.Done():
+			s.ticker.Stop()
+			return nil
+		}
+	}
+}
+
+// Update implements [service.Service] and applies settings.
+func (s *Service) Update(newConfig any) error {
+	newArgs := newConfig.(Arguments)
+
+	// We either never set the block on the first place, or recently removed
+	// it. Make sure we stop everything gracefully before returning.
+	if newArgs.URL == "" {
+		s.mut.Lock()
+		s.ch = nil
+		s.ticker.Reset(math.MaxInt64)
+		s.asClient = noopClient{}
+		s.args.HTTPClientConfig = config.CloneDefaultHTTPClientConfig()
+		s.mut.Unlock()
+		return nil
+	}
+
+	s.mut.Lock()
+	hash, err := newArgs.Hash()
+	if err != nil {
+		return err
+	}
+	s.dataPath = filepath.Join(s.opts.StoragePath, ServiceName, hash)
```

Comment:
I think we _do_ need to add more than just the URL for the hash location though, eg. in case the username changes we'll want to not try to load configuration from the previous user.

Do you suggest just manually hash the fields instead of the entire struct in that helper instead?

Reply:
Since this service controls the folder `[storage_path]/remotecfg` do we need to use a hash at all for where the config is cached? Or can we avoid that altogether?

(This is not to say we shouldn't hash the config contents and use that for the reloading and the caching logic)

Reply:
To be more specific, what are the downsides of storing the cached config at the full path `[storage_path]/remotecfg/cache.river`

Reply:
The downside is potentially running with a new username/password, failing to authenticate and then falling back to the cached configuration that already existed for another user, or another URL.